### PR TITLE
Fix a few cases where fill drag behaviour didn't work as expected

### PR DESCRIPTION
**Pull request by MrStevns**
- Filling on layer below when reference layer was all layers, would not allow stroke filling when applying on non transparent color
- Filling with overlay would result in multiple undo operations for one paint operation...
- Click fill with tablet would result in two fill events.

### CODE REVIEW

**Comment by MrStevns** [comment]
It would be nice to have this reviewed as it's also contains some fixes related to tablet behavior. I still believe that the implementation works as intended and there's tests to back it up but of course we should agree on the changes nevertheless

**Comment by chchwy** [comment]
Will be looking at this PR soon.

**Comment by chchwy** [comment]
I would suggest making another PR for the mouse/table event things next time. Let's focus on one thing at a time.

**Comment by MrStevns** [comment]
Noted 👍 . At the time it made sense for me because the tablet bug was reproducible and somewhat related to the drag behavior but yes in hindsight I should have done that

**Comment by MrStevns** [comment]
Any news regarding this PR, would it make it easier to review if I moved the double click event logic out?

File: core_lib/src/graphics/bitmap/bitmapbucket.cpp [hunk 1]
```diff
@@ -101,17 +105,21 @@ bool BitmapBucket::shouldFill(QPointF checkPoint) const
...
     }
 
+    QRgb fillToColor = mFillToImageColor;
+
     // Ensure that when dragging that we're only filling on either transparent or same color
-    if ((mReferenceColor == targetPixelColor && targetPixelColor == pixelColor) || (pixelColor == 0 && targetPixelColor == 0))
-    {
-        return true;
+
+    if (mFillToLayerPixelFormat.premultiplied() == QPixelFormat::Premultiplied) {
```
[the remark this answers is not in the document]
Comment (MrStevns) [reply core_lib/src/graphics/bitmap/bitmapbucket.cpp:110]: You're right, the tests works now without it \o/  I'll change it.

File: core_lib/src/graphics/bitmap/bitmapbucket.cpp [hunk 2]
```diff
@@ -101,17 +105,21 @@ bool BitmapBucket::shouldFill(QPointF checkPoint) const
...
-    // doesn't match the target color, otherwise it will fill the same color for no reason.
-    // We still expect to only fill on either transparent or same color.
-    if (mAppliedColor != targetPixelColor && pixelColor == 0)
-    {
+    if (mReferenceLayerPixelFormat.premultiplied() == QPixelFormat::Premultiplied) {
+        colorOfReferenceImage = qUnpremultiply(colorOfReferenceImage);
+    }
+
+    if (!BitmapImage::compareColor(targetPixelColor, mAppliedColor, mTolerance, mPixelCache) &&
+        BitmapImage::compareColor(targetPixelColor, mFillToImageColor, mTolerance, mPixelCache) &&
```
[the remark this answers is not in the document]
Comment (MrStevns) [reply core_lib/src/graphics/bitmap/bitmapbucket.cpp:119]: Good point, I will address it 👍

File: core_lib/src/graphics/bitmap/bitmapbucket.cpp [hunk 3]
```diff
@@ -101,17 +99,20 @@ bool BitmapBucket::shouldFill(QPointF checkPoint) const
...
-    {
-        return true;
+    QRgb fillToColor = mFillToImageColor;
+
+    // Using either of these modes applies premultiplied colors, so we have to unpremultiply to compare the colors
+    if (mProperties.bucketFillReferenceMode == 1) {
+        colorOfReferenceImage = qUnpremultiply(colorOfReferenceImage);
+    }
+    if (mProperties.bucketFillToLayerMode == 1) {
+        fillToColor = qUnpremultiply(fillToColor);
```
[the remark this answers is not in the document]
Comment (MrStevns) [reply core_lib/src/graphics/bitmap/bitmapbucket.cpp]: This was caused by an unrelated bug and has been fixed, so the check is no longer relevant

File: core_lib/src/graphics/bitmap/bitmapbucket.cpp [hunk 4]
```diff
@@ -84,12 +84,10 @@ bool BitmapBucket::shouldFill(QPointF checkPoint) const
...
     if (!targetImage.isLoaded()) { return false; }
 
-    BitmapImage referenceImage = mReferenceImage;
-
-    QRgb pixelColor = referenceImage.constScanLine(point.x(), point.y());
+    QRgb colorOfReferenceImage = mReferenceImage.constScanLine(point.x(), point.y());
     QRgb targetPixelColor = targetImage.constScanLine(point.x(), point.y());
 
-    if (mProperties.fillMode == 2 && pixelColor != 0)
+    if (mProperties.fillMode == 2 && colorOfReferenceImage != 0)
```
[the remark this answers is not in the document]
Comment (MrStevns) [reply core_lib/src/graphics/bitmap/bitmapbucket.cpp:92]: Fixed

File: core_lib/src/interface/scribblearea.cpp [hunk 5]
```diff
@@ -702,6 +706,22 @@ void ScribbleArea::handleDoubleClick()
...
 
+void ScribbleArea::tabletReleaseEventFired()
+{
+    // Under certain circumstances a mouse press event will fire after a tablet release event.
+    // This causes unexpected behaviours for some of the tools, eg. the bucket.
+    // The problem only seems to occur on windows and only when tapping.
+    // prior to this fix, the event queue would look like this:
+    // eg: TabletPress -> TabletRelease -> MousePress
+    // The following will filter mouse events created after a tablet release event.
+    mTabletReleaseMillisAgo += 100;
```
[the remark this answers is not in the document]
Comment (MrStevns) [reply core_lib/src/interface/scribblearea.cpp]: Fixed

File: core_lib/src/graphics/bitmap/bitmapbucket.cpp [hunk 6]
```diff
@@ -101,17 +105,21 @@ bool BitmapBucket::shouldFill(QPointF checkPoint) const
...
-    // We still expect to only fill on either transparent or same color.
-    if (mAppliedColor != targetPixelColor && pixelColor == 0)
-    {
+    if (mReferenceLayerPixelFormat.premultiplied() == QPixelFormat::Premultiplied) {
+        colorOfReferenceImage = qUnpremultiply(colorOfReferenceImage);
+    }
+
+    if (!BitmapImage::compareColor(targetPixelColor, mAppliedColor, mTolerance, mPixelCache) &&
+        BitmapImage::compareColor(targetPixelColor, mFillToImageColor, mTolerance, mPixelCache) &&
+        (BitmapImage::compareColor(colorOfReferenceImage, fillToColor, mTolerance, mPixelCache) || colorOfReferenceImage == 0)) {
```
[the remark this answers is not in the document]
Comment (MrStevns) [reply core_lib/src/graphics/bitmap/bitmapbucket.cpp]: Fixed

**Comment by MrStevns** [comment]
PR should be ready for review again

**Comment by davidlamhauge** [comment]
Basically, I like the fill tool. I like the expand fill, and I also like the possibility to fill on the layer below.
On the other hand, I must side with Jakob, that there are many combinations, and it can be confusing.
A week ago I myself ruined a background for a project, because I accidently made a fill on the layer below, and when I realised what I'd done, the background was beyond repair. It was 100% my fault, but I'd wish there was some warning or indication, that I was about to fill on another layer. I will probably not be the only one making that mistake.
Besides this bad experience, I am very pleased with the fill tool.

File: core_lib/src/graphics/bitmap/bitmapbucket.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (MrStevns) [reply core_lib/src/graphics/bitmap/bitmapbucket.cpp]: Hmm I must have forgotten it or discarded the removal during my testing.. it's gone now.

**Comment by MrStevns** [comment]
I feel like talking about how the feature should behave is a little too late or not relevant as far as this PR goes.. the functionality is already there, this PR was only meant to fix bugs regarding existing functionality, not reinvent or remove the existing implementation. Can't we take this discussion in some other way that doesn't involve stalling the PR or derailing its purpose?

@J5lx 
As far as the tool goes, you don't have to use all its features, just like blending modes, each thing is a tool one can use for a certain task. Of course if you feel like you're developing a distaste for the tool because it's getting too complex, then we should look into that.

**Comment by MrStevns** [comment]
> I’m very sorry. I didn’t want to make you feel like your work was being artificially held up or sidetracked. I need to be more careful with my words.

It's hard to convey emotions through text, and depending on what time of the day one reads and answers something, intend and understanding may differ. No need to apologize 😅 

> The entire remark about the complexity of the fill tool has nothing to do with the changes introduced by this PR and is just an expression of my frustration at the unique challenges of reviewing them. Not because these particular changes are especially complex or anything, but because the fill tool in general, with its various parameters and the many possible combinations of them, is complex per se (in regard to its inner workings, not so much in regard to its usefulness in various workflows). And I’m not even sure if there’s much that can be done about it. That is, I don’t necessarily think it’s too complex, I just think it’s complex.

That one is mostly on me, I should have written down the functionality the first time I made this proposal, so it would be clear how I intended it to work and so that in the future we would be to read up on its functionality. Hopefully the following write up makes it clearer 😄 

> The reason I was still talking about intended functionality is not because I want to change it or turn this PR into something else, but because I feel that I should understand the intended functionality in order to properly assess whether this code implements it and I’m unsure if my current understanding of it is correct. More specifically, the current behavior (of this PR) didn’t feel right for me and I wasn’t sure if that was because there’s actually something wrong with it or just because I misunderstood the intention behind it. The behavior has changed throughout this PR in response to feedback and I wanted to make sure that all of it is still part of the fixes.

That also makes sense, it's difficult to review something if it's too complex to understand from code alone or simply has too many combinations to understand the intend. 
It's true that the behavior has changed a few times during the lifecycle of the PR, either to address a bug or me realizing an odd behavior.. so it's a good thing to have this clarified once a for all and not basing it on loose assumptions or what I say and don't say, is right.

## Understanding the feature
The idea behind the drag to fill behavior is that you should be able to apply several fills in easy and less error prone manner than manually clicking on a pixel several times, and we can do that because we know where you started. 
To keep things simpler, we should be able to apply the following rules in all cases:

1. You should be able to drag across and not fill any other pixel than transparent when the start drag position is a transparent pixel.
2. When the start drag position is a colored pixel, only a pixel of similar (based on tolerance) color or an area which is transparent will be filled.

#### How I tested:
Original image:
image
project: [fill-drag-pr.pclx.zip](https://github.com/pencil2d/pencil/files/9983275/fill-drag-pr.pclx.zip)

#### Bucket settings:
Blend mode: replace
Reference mode: current layer

[master1]: https://user-images.githubusercontent.com/1045397/200907802-145f7b6b-d993-431c-b35d-d2a2c181926b.gif
[pr1]: https://user-images.githubusercontent.com/1045397/200912773-1ac56eb8-18d5-4304-9110-eda48a1e7c86.gif
[master2]: https://user-images.githubusercontent.com/1045397/200910411-4e93b02d-811b-4450-9638-6562dc35ec09.gif
[pr2]: https://user-images.githubusercontent.com/1045397/200913721-bf58103c-e42a-4288-b2b6-dc785437a909.gif

### Filling on the same layer

Rule |       Master         |             PR               |           Functionality       |           Conclusion
:---:|:----------------:|:-------------------:|:----------------------:|:---------------------   
You should be able to drag across and not fill any other pixel than transparent when the start drag position is a transparent pixel. | ![master1] | ![pr1] |  same | Works as intended
When the start drag position is a colored pixel, only a pixel of similar (based on tolerance) color or an area which is transparent will be filled.: | ![master2] | ![pr2] | same | Works as intended


[master3]: https://user-images.githubusercontent.com/1045397/201159703-bdb15b1f-726c-450d-9af8-c0de020beaf3.gif
[pr3]: https://user-images.githubusercontent.com/1045397/201160027-b443946a-084a-4c20-b120-7d8bacb6ea7d.gif
[master4]: https://user-images.githubusercontent.com/1045397/201161815-5f770c30-2724-4d3e-bbd8-49c5ce21846f.gif
[pr4]: https://user-images.githubusercontent.com/1045397/201162651-b5339fb1-2051-435b-b42c-36050547ebf7.gif

### Filling on layer below

Rule |       Master         |             PR               |           Functionality        |        Conclusion
:---:|:----------------:|:-------------------:|:-----------------------:|:------------------- 
You should be able to drag across and not fill any other pixel than transparent when the start drag position is a transparent pixel. | ![master3] | ![pr3] |  Differ, on master the target layer avoids filling where the reference color is, even though it's a different layer. On the PR branch, we don't take this into account anymore. |     PR algorithm is more loose about where to fill, is this better or worse? I'd say worse.. I think the master implementation is better.
When the start drag position is a colored pixel, only a pixel of similar (based on tolerance) color or an area which is transparent will be filled.: | ![master4] | ![pr4] | Differs, on master the start pixel is filled and all other colors that are not the same or transparent are ignored. In the PR, the white background is kept, and filling happens on all reference points, because the fill to layer is blank. | PR algorithm doesn't work entirely, the white pixels should have been filled and as with the Rule1, we should try to be smarter about where we paint, even if we're working on a different layer.

In other words... there's still code to be written in order to make sure the rules apply, when painting on the layer below.

**Comment by MrStevns** [comment]
After having tried a few things, I'm pretty sure i have the algorithm sorted out now... now the only problem is that I don't like the fillTo feature anymore 😆

The reason is that when it's correctly implemented (ie. not the result of master...) then it's only usable when the reference is all layers, at least if you consider the rules I wrote down in the previous post. For example take test project: 

filldrag-fillTo-stuff

I start by filling the layer below with a green color, at this point the reference is set to 'all layers', and both reference and target pixel is transparent, thus it works. Next I change the color to Purple and fill again, as I try it out with reference: 'all layers', it continues to work fine because we're scanning all layers, because both the reference and target color is found. This however doesn't work when I afterwards change to reference: 'current layer', because in this case the reference pixel is still transparent, only the target pixel is green and therefore it doesn't fill.

I thought it was a pretty cool feature in the beginning, because you would be able to both draw the lines of your animation and do the filling efficiently after, without having to jump between layers but the additional complexity it adds to the drag algorithm, makes me think otherwise.

**Comment by MrStevns** [comment]
Thanks for reviewing again Jacob, I've made a few smaller changes to make up for the more limited functionality of the fillTo layer mechanism now and fixed a bunch of typos. I have also merged your PR, so I'd say it's ready to be reviewed again
